### ISSUE TITLE
Changes that require new libraries to test cause build failures

### ISSUE THREAD

**Issue by Kirill888** [opendatacube/odc-tools#318]
Problem
======

You add new code that uses new library, you write tests, you update test docker to have new library, your tests pass locally. But when you make PR old test docker image is used to test your code and your tests fail on CI.

Proposed Solution
==============

Ideally one would always build docker image as described by the current branch to run the tests, but that can be slow and might also fail for random external reasons that have nothing to do with the current changes. I propose we implement a more complex workflow where we 

1. Check if feature branch has changes that are likely to require new docker image (files changed under `docker/**` or any `setup.cfg` files when diffing with `develop` branch)
   - YES: build fresh docker from this branch
   - NO: just pull in default test docker (saves time and reduces probability of dealing with transient issues)

And if running on `develop` branch (after PR is merged), similar rules can apply but instead of diffing branches you diff files changed since the last build, github provides enough information to figure this out programmatically.

So basically:
- Check if docker rebuild is needed based on changes being tested
   - `[YES]` Build fresh docker
   - `[NO ]` Pull "cached" docker from dockerhub
